CE3 + http4s 0.23.x upgrade

andreamarcolin opened the pull request:
Follow up to #172 

This PR upgrades cats-effect to 3.x, on both http4s-emitter and metadata modules, which also means upgrading to fs2 3.
Http4s was also updated to 0.23.x the most recent stable release built on top of CE3 and fs2 3.

A few things changed in these major versions, directly impacting http4s-emitter code.

First, `fs2.concurrent.Queue` has been reworked and moved into cats-effect (`cats.effect.std.Queue`). `Enqueue` and `Dequeue` (fs2 parent traits of `Queue`) have been renamed to `QueueSource` and `QueueSink`. Note that, confusingly enough, `Dequeue` in cats-effect in an entirely different data structure, i.e. a double-ended queue.
`enqueue1` has been basically renamed to `offer` (blocking semantics when queue is full), while `offer1` has been renamed to `tryOffer` (immediately returns with a boolean signaling success/failure).
See https://github.com/typelevel/fs2/blob/v2.5.10/core/shared/src/main/scala/fs2/concurrent/Queue.scala and https://github.com/typelevel/cats-effect/blob/series/3.x/std/shared/src/main/scala/cats/effect/std/Queue.scala 
`Queue.dequeue` has been dropped, and replaced by a bunch of `Stream.fromQueue...` methods. The old `dequeue` used to return a stream, and dequeuing items in chunks as big as possible by default (limit = Int.MaxValue), see 
https://github.com/typelevel/fs2/blob/24282e2268f489972252433b4b85e8c8ecf5109b/core/shared/src/main/scala/fs2/concurrent/Queue.scala#L84. The same API in fs2 3 is implemented by https://github.com/typelevel/fs2/blob/6b4be6642e8d83f2308d99a41b66796338b3d147/core/shared/src/main/scala/fs2/Stream.scala#L3339

Second, CE3's typeclass hierarchy is vastly different from the one in CE2, almost reversed. Migration was easy enough, with `Timer` becoming `Temporal`. `Temporal` now extends `Concurrent`, and in a few places, requiring Temporal + Sync/Concurrent was causing conflicts in the implicit scope, due to both extending the same parent traits (e.g. Monad, MonadError). So I found out that the idea is just to replace them with the least powerful typeclass that supports both constraint, which in many cases ends up being `Async`. This shouldn't be a problem for end users though. See https://gitter.im/typelevel/cats-effect?at=603ddd125d0bfb4e587f5d14

Third and final point, CE3 comes with UUIDGen and Random typeclasses, so that there is no need to use java methods wrapped in `Sync.delay`. This also addresses #154

Target of this PR is `release/2.0.0`, since this upgrade contains breaking changes. MiMa previous versions already cleared before merging #172 to the release branch.

snowplowcla commented:
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

andreamarcolin commented:
Very interesting test failures I run into, tests that were passing fine on my Mac 100% of the times.

The first one was on `MetadataSpec`: the timeout test was returning the actual response after a 5 seconds sleep even if the action was supposedly being applied a 3 seconds timeout (`Concurrent[F].timeoutTo`). 
After thinking about it, I thought it had to do with the explicit `Thread.sleep` in the test, so the hint was that the sleep done in 2 tests (executed in parallel by specs2) were probably blocking all CE's IORuntime compute pool threads. 
I managed to reproduce locally by passing -XX:ActiveProcessorCount=1 as a VM argument, simulating a machine with one physical thread of execution (1 vCPU, as I assume GH runners have 1 vCPU). That indeed confirmed my hypothesis, IORuntime's compute pool [gets instantiated with 2 threads](https://github.com/typelevel/cats-effect/blob/71110ecd96b42453456063576a46bffa56f89d95/core/jvm/src/main/scala/cats/effect/unsafe/IORuntimeCompanionPlatform.scala#L36) and both end up being blocked by "Thread.sleep", so the timeout action doesn't even have a chance to be scheduled. 
The solution was to change the `Sync[F].delay` that wraps the client call to `Sync[F].interruptible`, which both shifts the computation to the blocking pool (unbounded), and also tries to interrupt the action that timed out, using the underlying `Thread.interrupt` on the JVM. This both made the timeout return the correct result, and cleans up resources if timeout kicks in, avoiding wasting time in case of `Thread.sleep`.

Second failing test was on `Http4sEmitterSpec`: tests were using a mix of IO, AtomicInteger and the `eventually` assertion, that sleeps many times until eventually the assertion passes. All this gave me a headache, since tests were passing without `-XX:ActiveProcessorCount=1`, while just one was failing with the VM option, but only when the whole Spec was run. If the broken case was run alone, it would pass. So again, the Thread.sleep across the suite and assertions were probably giving a hard time to the IORuntime. 
What I decided to do here since the problem was probably in the spec was to rewrite it in purely functional style, with `Ref` and `IO.sleep` to check the number of requests sent by the emitter after buffering, both before and after the emitter gets shutdown and queue is flushed.

istreeter reviewed (commented):
Hey @andreamarcolin this looks great as far as I can tell!  Thank you very much for your work on this.

Having said that.... I am not yet fully familiar with the Cats Effect 3 syntax, so I might not be the best person to review this.  I think @oguzhanunlu has spent more time learning about CE3 so he can probably give you a better review

oguzhanunlu reviewed (approved):
Hey @andreamarcolin , very interesting outcomes here and thanks for the detailed explanation! Overall it looks good to me from a functionality point of view but I have a comment about not limiting the exposure of `Async`/`Sync`.

Unlike CE2, `Async`/`Sync` are the most powerful typeclasses in CE3 and it is recommended to limit their exposure. When we have `Async` constraint in a method, it will propagate that constraint through the codebase which in turn will reduce local reasoning. Dependent parametric methods' signatures will become less reasonable since they will all have `Async`. A recommended way to limit their exposure in signatures is to make use of capability traits so that we can provide better parametric reasoning and limit what that method can do to a discrete set of operations. An example of this approach is `fs2.io.file.Files[F]` capability trait in FS2 3.x

Feel free to ignore my comment, it is definitely not a blocker for the release.

oguzhanunlu commented on modules/http4s-emitter/src/main/scala/com/snowplowanalytics/snowplow/scalatracker/emitters/http4s/Http4sEmitter.scala:
```diff
@@ -85,23 +82,23 @@ object Http4sEmitter {
-  private def shutdown[F[_]: Concurrent: Timer](
-    fiber: Fiber[F, Unit],
-    queue: Enqueue[F, Action],
+  private def shutdown[F[_]: Concurrent: Temporal](
```
just a nit, Temporal extends Concurrent

oguzhanunlu commented on modules/http4s-emitter/src/main/scala/com/snowplowanalytics/snowplow/scalatracker/emitters/http4s/Http4sEmitter.scala:
```diff
@@ -85,23 +82,23 @@ object Http4sEmitter {
-  private def shutdownWithTimeout[F[_]: Concurrent: Timer](
-    fiber: Fiber[F, Unit],
-    queue: Enqueue[F, Action],
+  private def shutdownWithTimeout[F[_]: Concurrent: Temporal](
```
same as https://github.com/snowplow/snowplow-scala-tracker/pull/173/files#r813928299

andreamarcolin commented:
Hey @oguzhanunlu! Thanks for the review, I already addressed the two redundant context bounds. 

When it comes to `Sync/Async` I definitely agree with you and I am aware of the implications of those two. On the other hand, I didn't see a quick way out of it. `Async` in particular is an unfortunate consequence of `Http4sEmitter.sendRequest` requiring both `Sync` for `invokeCallback` and `Temporal` for sleeping between retries.
As I mentioned in the long explanation, this seems to be a CE limitation, and I addressed it as suggested by Daniel (Spiewak) here https://gitter.im/typelevel/cats-effect?at=603ddd125d0bfb4e587f5d14.
I even tried to keep just Sync + Temporal and pass implicits explicitly to functions, but that didn't work either (syntax doesn't work due to ambiguous implicits).

The same can be said for `GceMetadata` and `Ec2Metadata`: they both required `Sync` (no way out of it, the code wraps a potentially blocking operation that is passed by the end user, and that could even launch missiles), but also `Concurrent` for timeouts, and the Sync+Concurrent combo suffers of the same problems mentioned above.

I am not sure where you would like to see capability traits. The only one I can think about is to replace the `Sync` constraint in `Http4sEmitter` with a log4cats `Logger`, given that the only reason Sync is needed is for putting a warn log in `invokeCallback`. I wasn't sure about introducing a new library though

oguzhanunlu commented:
> I didn't see a quick way out of it.

I also didn't see a quick way for the most part without at least partial refactoring and it'd be an overengineering for the remaining small scope. e.g. for metadata package, we could move parametricity from class def to method defs and restrict what methods could do further -- now we can't really restrict the classes' `F` without defining an unnecessary layer of DSL on top of CE.

Apart from this release, to avoid requiring `Async`/`Sync` constraints in general for CE3 apps, we can make use of capability traits with `Async`/`Sync` instances, and use that trait in context bound to prevent arbitrary side effects & provide better reasoning.

I'd say let's proceed with what we have without further effort and keep this strategy in mind -- assuming you also would like to stay away from `Async`/`Sync` constraints in context bounds in any non-trivial CE3 app.

Let me add that I really enjoyed reading your comments and benefit from your PR explanation around fs2 3's queue updates, so thank you again for the detailed comments incl. the ref to gitter. 👍

chuwy reviewed (approved):
The PR has been seating in by inbox for a while and it looks great!

On the following question:

> The only one I can think about is to replace the Sync constraint in Http4sEmitter with a log4cats Logger, given that the only reason Sync is needed is for putting a warn log in invokeCallback. I wasn't sure about introducing a new library though

I certainly don't mind to introduce log4cats dependency, I think it's a standard de-facto for pretty much any our lib (just if we use it - let's make sure it prints it with tracker's namespaces, not original app's `Main` class, which is the case if you use `Logging` context bound)
